File: bumperbot_ws/src/bumperbot_localization/bumperbot_localization/kalman_filter.py
#!/usr/bin/env python3
# =============================================================================
# kalman_filter.py  —  bumperbot_localization
# =============================================================================
# ROS 2 Python node implementing a 1-D Kalman filter for angular velocity.
# Python variant — the C++ version (kalman_filter.cpp) is also available.
#
# Purpose:
#   Fuse noisy wheel-encoder angular velocity with IMU gyroscope angular
#   velocity to produce a cleaner estimate of the robot's yaw rate.
#
# Inputs:
#   /bumperbot_controller/odom_noisy  (Odometry)  — from noisy_controller
#   /imu/out                          (Imu)        — from mpu6050_driver
#
# Output:
#   /bumperbot_controller/odom_kalman (Odometry)  — same as odom_noisy but
#   with twist.twist.angular.z replaced by the Kalman-filtered yaw rate.
#
# Algorithm (1-D Gaussian belief tracking):
#   1. State Prediction  — shift mean by motion increment, increase variance
#      by motion noise to reflect uncertainty growth over time.
#   2. Measurement Update — blend IMU measurement with predicted state;
#      trusted more when sensor variance < state variance.
# =============================================================================

import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
from sensor_msgs.msg import Imu


class KalmanFilter(Node):

    def __init__(self):
        super().__init__("kalman_filter")

        # Subscribe to the noisy wheel-encoder odometry
        self.odom_sub_ = self.create_subscription(Odometry, "bumperbot_controller/odom_noisy", self.odomCallback, 10)
        # Subscribe to raw IMU — we only use angular_velocity.z (yaw rate)
        self.imu_sub_ = self.create_subscription(Imu, "imu/out", self.imuCallback, 10)
        # Publish the filtered odometry (all fields copied from odom_noisy,
        # angular.z replaced by the Kalman estimate)
        self.odom_pub_ = self.create_publisher(Odometry, "bumperbot_controller/odom_kalman", 10)
        
        # Kalman state: Gaussian belief over angular velocity
        self.mean_ = 0.0       # current best estimate of angular velocity (rad/s)
        self.variance_ = 1000.0  # initially very uncertain — let measurements dominate

        # Noise parameters — tuned to trust IMU more than wheel odometry
        self.motion_variance_ = 4.0      # how much uncertainty grows per prediction step
        self.measurement_variance_ = 0.5 # how noisy the IMU measurement is (lower = more trusted)

        # Latest IMU angular velocity reading (updated asynchronously)
        self.imu_angular_z_ = 0.0

        # Skip the first odometry message — we need two to compute a delta
        self.is_first_odom_ = True
        self.last_angular_z_ = 0.0  # angular velocity from the previous odom message
        self.motion_ = 0.0          # change in angular velocity between odom callbacks

        # Output message; copied from odom_noisy with angular.z overwritten
        self.kalman_odom_ = Odometry()


    def odomCallback(self, odom):
        """Main filter loop — called each time a new noisy odometry arrives."""
        self.kalman_odom_ = odom  # copy all pose and velocity fields

        if self.is_first_odom_:
            # Bootstrap: initialise the filter state from the first measurement
            self.last_angular_z_ = odom.twist.twist.angular.z
            self.is_first_odom_ = False
            self.mean_ = odom.twist.twist.angular.z
            return
        
        # Compute how much the angular velocity changed since the last callback
        self.motion_ = odom.twist.twist.angular.z - self.last_angular_z_

        # Step 1: predict new state based on motion (uncertainty grows)
        self.statePrediction()
        # Step 2: correct prediction using the IMU measurement
        self.measurementUpdate()

        # Save current angular velocity as baseline for next delta
        self.last_angular_z_ = odom.twist.twist.angular.z

        # Replace the noisy angular velocity with the filtered estimate
        self.kalman_odom_.twist.twist.angular.z = self.mean_
        self.odom_pub_.publish(self.kalman_odom_)


    def imuCallback(self, imu):
        """Cache the latest IMU yaw-rate measurement for use in measurementUpdate."""
        self.imu_angular_z_ = imu.angular_velocity.z


    def measurementUpdate(self):
        """Correct the predicted state using the IMU measurement (Bayes update).

        Kalman measurement update equations:
          mean     = (σ_meas * μ + σ_state * z) / (σ_state + σ_meas)
          variance = (σ_state * σ_meas)         / (σ_state + σ_meas)

        A low measurement_variance_ (trusted sensor) pulls the mean strongly
        toward the IMU reading and shrinks the state variance.
        """
        self.mean_ = (self.measurement_variance_ * self.mean_ + self.variance_ * self.imu_angular_z_) \
                   / (self.variance_ + self.measurement_variance_)
                     
        self.variance_ = (self.variance_ * self.measurement_variance_) \
                       / (self.variance_ + self.measurement_variance_)


    def statePrediction(self):
        """Advance the state estimate using the wheel-encoder motion model.

        Kalman state prediction equations:
          mean     = mean + motion
          variance = variance + motion_variance

        Adding motion_variance_ increases the uncertainty because encoder
        readings are noisy and we can't perfectly predict the new angular rate.
        """
        self.mean_ = self.mean_ + self.motion_
        self.variance_ = self.variance_ + self.motion_variance_


def main():
    rclpy.init()

    kalman_filter = KalmanFilter()
    rclpy.spin(kalman_filter)
    
    kalman_filter.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: bumperbot_ws/src/bumperbot_localization/src/kalman_filter.cpp
// =============================================================================
// kalman_filter.cpp  —  bumperbot_localization
// =============================================================================
// C++ implementation of a 1-D Kalman filter for angular velocity fusion.
// Selected by default (use_python:=False) in local_localization.launch.py.
// Python equivalent: bumperbot_localization/kalman_filter.py
//
// Fuses two sources of angular velocity (yaw rate) information:
//   - Wheel encoder odometry (/bumperbot_controller/odom_noisy) — motion model
//   - IMU gyroscope (/imu/out)                                  — measurement
//
// See kalman_filter.hpp for the full algorithm and parameter description.
// =============================================================================

#include "bumperbot_localization/kalman_filter.hpp"

using std::placeholders::_1;

KalmanFilter::KalmanFilter(const std::string &name)
    : Node(name), mean_(0.0), // initial angular velocity estimate: 0 rad/s
      variance_(
          1000.0), // high initial uncertainty → let first measurement dominate
      motion_variance_(4.0), // process noise (constant — tunable)
      measurement_variance_(
          0.5), // IMU noise (lower = more trusted than encoder)
      motion_(0.0),
      is_first_odom_(true), // bootstrap flag — skip first callback
      last_angular_z_(0.0), imu_angular_z_(0.0) {
  // Subscribe to noisy encoder odometry — drives the prediction step
  odom_sub_ = create_subscription<nav_msgs::msg::Odometry>(
      "bumperbot_controller/odom_noisy", 10,
      std::bind(&KalmanFilter::odomCallback, this, _1));
  // Subscribe to raw IMU — cache for the measurement update step
  // Large queue (1000) to tolerate IMU arriving faster than odom
  imu_sub_ = create_subscription<sensor_msgs::msg::Imu>(
      "imu/out", 1000, std::bind(&KalmanFilter::imuCallback, this, _1));
  // Publish filtered odometry — same as odom_noisy but angular.z replaced
  odom_pub_ = create_publisher<nav_msgs::msg::Odometry>(
      "bumperbot_controller/odom_kalman", 10);
}

void KalmanFilter::odomCallback(const nav_msgs::msg::Odometry &odom) {
  // Copy the full odom message — we'll overwrite only angular.z at the end
  kalman_odom_ = odom;

  // Bootstrap: on the very first message, initialise the filter state and
  // return
  if (is_first_odom_) {
    last_angular_z_ = odom.twist.twist.angular.z;
    mean_ =
        odom.twist.twist.angular.z; // seed the filter with the first reading
    is_first_odom_ = false;
    return;
  }

  // Motion = change in angular velocity since the last odom callback.
  // This drives the prediction step (the "u" term in the filter).
  motion_ = odom.twist.twist.angular.z - last_angular_z_;

  // Run one full Kalman cycle: predict then correct
  statePrediction();
  measurementUpdate();

  // Save current reading as baseline for the next delta
  last_angular_z_ = odom.twist.twist.angular.z;

  // Publish the corrected odometry with the filtered angular.z
  kalman_odom_.twist.twist.angular.z = mean_;
  odom_pub_->publish(kalman_odom_);
}

void KalmanFilter::imuCallback(const sensor_msgs::msg::Imu &imu) {
  // Cache the latest IMU yaw rate for use in measurementUpdate().
  // The IMU publishes at a different rate than odom, so we just store
  // the most recent value and use it on the next odom callback.
  imu_angular_z_ = imu.angular_velocity.z;
}

void KalmanFilter::measurementUpdate() {
  // Bayes update — blend the prediction with the IMU measurement.
  // mean  = (σ_sensor · μ_pred + σ_pred · z_imu) / (σ_pred + σ_sensor)
  // var   = (σ_pred · σ_sensor)                  / (σ_pred + σ_sensor)
  //
  // The IMU (σ=0.5) is trusted more than the encoder model (σ can grow large),
  // so the filtered mean is pulled toward imu_angular_z_ each cycle.
  mean_ = (measurement_variance_ * mean_ + variance_ * imu_angular_z_) /
          (variance_ + measurement_variance_);

  variance_ =
      (variance_ * measurement_variance_) / (variance_ + measurement_variance_);
}

void KalmanFilter::statePrediction() {
  // Apply the motion model: advance the mean by the encoder-derived motion
  // and inflate the variance to reflect growing uncertainty.
  // Adding motion_variance_ models the fact that wheel odometry is imperfect.
  mean_ = mean_ + motion_;
  variance_ = variance_ + motion_variance_;
}

int main(int argc, char *argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<KalmanFilter>("kalman_filter");
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}